### ISSUE TITLE
Blazor AOT .Net6 Preview 7 - Still downloading Dlls

### ISSUE THREAD

**Issue by srpeirce** [dotnet/aspnetcore#35302]
### Describe the bug
I raised this issue previously, but I cannot find anywhere where it is being tracked to see status (apologies if I'm just missing it - happy to be linked to another ticket).

https://github.com/dotnet/aspnetcore/issues/33730

There are still a large amount of `dlls` being downloaded with AOT. Perhaps it is my misunderstanding of how it works but I do not expect the dlls, just a larger `.wasm` file?

Larger wasm file than before AOT:
![image](https://user-images.githubusercontent.com/12847647/129261284-b9423093-1710-42b7-bb0e-c19a89540d39.png)

But still a large number of dlls:
![image](https://user-images.githubusercontent.com/12847647/129261294-79247ca2-48c9-4f7b-9368-bdf6f7b36eaf.png)

### To Reproduce
Install workload: `dotnet workload install wasm-tools`

Project with `RunAOTCompilation` set to true.
```xml
<PropertyGroup>
    <TargetFramework>net6.0</TargetFramework>
    <ServiceWorkerAssetsManifest>service-worker-assets.js</ServiceWorkerAssetsManifest>
    <BlazorEnableTimeZoneSupport>false</BlazorEnableTimeZoneSupport>
    <BlazorWebAssemblyPreserveCollationData>false</BlazorWebAssemblyPreserveCollationData>
  </PropertyGroup>

  <PropertyGroup Condition="'$(Configuration)'!='DEBUG'">
    <!-- Bug in .Net 6 Preview 7 - this should not be non-debug only. https://github.com/dotnet/runtime/issues/55750 -->
    <InvariantGlobalization>true</InvariantGlobalization>

    <RunAOTCompilation>true</RunAOTCompilation>
  </PropertyGroup>
```

Publish application.

### Further technical details
```
.NET SDK (reflecting any global.json):
 Version:   6.0.100-preview.7.21379.14
 Commit:    22d70b47bc

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.19043
 OS Platform: Windows
 RID:         win10-x64
 Base Path:   C:\Program Files\dotnet\sdk\6.0.100-preview.7.21379.14\

Host (useful for support):
  Version: 6.0.0-preview.7.21377.19
  Commit:  91ba01788d
  ```

**Comment by srpeirce** [comment]
CC @javiercn @pranavkm

**Comment by pranavkm** [comment]
@radical is AOT expected to remove / trim out managed binaries?

**Comment by danroth27** [comment]
The DLLs are still expected. Even with AOT compilation, there are scenarios where we need to fall back to using the interpreter with the .NET binaries. This increases the download size, but provides a more compatible .NET experience.

**Comment by srpeirce** [comment]
Thanks for the clarification.

As a follow-up, I was expecting/hoping to see an initial speed boost on application start given the AOT performance compared to interpreted - I haven't done any formal metrics but I don't see any real difference (I have some stopwatches that output some numbers as a crude measure).

Is this to be expected? I had speculated that it might be related to the dlls still being downloaded and used?

**Comment by danroth27** [comment]
@srpeirce AOT doesn't improve the load time of the app. It actually makes load time worse because it makes the app size large. What AOT improves is the runtime performance of the app. Blazor WebAssembly in earlier releases uses a .NET IL interpreter implemented in WebAssembly for its runtime. While this interpreter is adequate for many scenarios, it isn't great for CPU intensive tasks. .NET on desktop and server uses JIT compilation to speed things up, but WebAssembly doesn't support JIT scenarios. AOT gives you the performance benefits of JIT compilation by compiling directly to WebAssembly up front. The app size increases because the representation of the high level .NET IL instructions in WebAssembly requires more WebAssembly instructions. So using AOT trades off load time performance for better runtime performance.

**Comment by mikes-gh** [comment]
@danroth27 What exactly is the dotnet runtime doing after it has been loaded by the browser?  The dotnet assets seem loaded quite quickly then a pause of 1.5 secs on a fast machine before the page starts to load the additional assets it needs.

![image](https://user-images.githubusercontent.com/16208742/129380485-da4207fc-bd81-472a-a25d-42fc46b80578.png)

**Comment by srpeirce** [comment]
Thanks @danroth27 for the detailed explanation, I appreciate your time 👍.

I'll have another play with AOT and decide if the tradeoffs are worth it for us. 

Have a great weekend! 🎉

**Comment by srpeirce** [comment]
Re-opened as I didn't spot @mikes-gh question for @danroth27. Apologies!

**Comment by Andrzej-W** [comment]
> The DLLs are still expected. Even with AOT compilation, there are scenarios where we need to fall back to using the interpreter with the .NET binaries. This increases the download size, but provides a more compatible .NET experience.

@danroth27 DLLs are blocked by firewalls (https://github.com/dotnet/aspnetcore/issues/31048). Is there any hope that we will be able to create WASM only Blazor application? I prefer to have a little less compatible .NET experience than useless application which cannot be downloaded by some potential users. What are those scenarios where you need to fall back to .NET interpreter?

**Comment by mikes-gh** [comment]
> Re-opened as I didn't spot @mikes-gh question for @danroth27. Apologies!

Thanks the reason I asked is because I am trying to optimise the startup time of our WASM app.  We managed to go from 24MB to 12MB initial payload just by moving to net6 which I presume is down to trimming but the load time wasn't much different.  Is there anything else I can look to optimise the wasm app startup time?

**Comment by SteveSandersonMS** [comment]
Moving to discussions. @danroth27 let us know if you need any info in this discussion.

**Comment by charlesroddie** [comment]
> The DLLs are still expected. Even with AOT compilation, there are scenarios where we need to fall back to using the interpreter with the .NET binaries. This increases the download size, but provides a more compatible .NET experience.

@danroth27 That's not AOT. Calling this AOT will make it harder to discuss AOT compilation support. There is an AOT issue (https://github.com/dotnet/aspnetcore/issues/5466) which is marked as completed, but if this means that IL is still being deployed to machines and interpreted then we will need another tracking issue for AOT support.

A linked AOT issue [Developers can AOT compile .NET apps into WebAssembly format](https://github.com/dotnet/runtime/issues/44316) says that it is full AOT that has been implemented, but what you are saying suggests that it is "mixed mode AOT" which has been implemented.

Is full AOT possible for developers to use or not?

**Comment by ivan-prodanov** [comment]
@charlesroddie It's not really mixed mode. I haven't tried latest .NET 6 yet, but in the alpha versions 9 months ago the AOT stripped off dlls of their method implementations and left only metadata (class names, method names) I guess probably to be forward compatible with reflection.

**Comment by srpeirce** [comment]
I don't think they are stripped dlls, they are same download size as interpreted.

**Comment by legistek** [comment]
> The DLLs are still expected. Even with AOT compilation, there are scenarios where we need to fall back to using the interpreter with the .NET binaries.

Dan are there scenarios besides reflection? 
And does this mean the WHOLE binary is included or just a pared down version with the metadata? I was almost certain I read that it was the latter.

**Comment by Andrzej-W** [comment]
I have made a simple test in VS 2022 Preview 3.1 and .NET 6 Preview 7. Default Blazor Wasm ASP.NET Core Hosted project (interpreted) after publication downloads 31 DLLs (total uncompressed size about 2MB) and dotnet.wasm file (also ~2MB uncompressed). With AOT enabled I have exactly the same DLL's but dotnet.wasm is much larger (~13MB uncompressed) and I think it now contains all the code compiled to WASM **and** the .NET interpreter.

I hope it will possible to remove all DLLs in the future and at the same time remove .NET interpreter. Instead of 32 files (total size ~4MB) we will have one wasm file which will be a little bit smaller (about ~11 MB uncompressed, ~3.5 MB compressed).

Why do we need DLLs and .NET interpreter? I think they are needed for JSON serialization/deserialization. They use reflection and dynamic code generation - both are impossible in WASM. If this is the only reason then maybe we can remove the interpreter and DLLs:
https://devblogs.microsoft.com/dotnet/try-the-new-system-text-json-source-generator/

Of course for those who need reflection there should be an option to include interpreter.

**Comment by legistek** [comment]
The interpreter shouldn't actually be needed just for reflection. Reflection basically just needs a database keyed with type names and member names. That could be done with an SQLite database or even JavaScript code generators.  It shouldn't need the IL bytecode. 

If there is something else going on in "AOT" besides reflection that does actually use the bytecode then this is a very significant compromise. I hope @danroth27 or others can provide some clarity. Thanks!

**Comment by jeromelaban** [comment]
Current builds of .NET 6 AOT are forced into mixed (AOT+Interpreter) mode because of some restrictions (https://github.com/dotnet/runtime/issues/56309) and even if FullAOT (no interpreter) was supported (which cannot be because of the lack of exceptions support in the Wasm spec related to the .NET `try/catch when/finally` blocks), assemblies will still needed for reflection and many other internal runtime aspects.

Assembly stripping (body removal) is only possible when the code has been fully AOTed and at this point, it's not possible, hence why assemblies are still fully downloaded.

**Comment by kenchan97** [comment]
um......
disappointed to see this.
expecting Blazor can get some frontend market share from JS

**Comment by sorainnosia** [comment]
in .NET 6 WASM AOT, the dotnet.wasm becomes very big 30+ MB in my case, while still keeping the normal DLLs which means that wasm AOT is doubling the download size compare to non AOT version, because the AOT's dotnet.wasm contains the wasm of the DLLs while still keeping a separate copy of those DLLs and yet required to be downloaded by the browser too, it makes no sense to have wasm version of the DLLs combine into gigantic dotnet.wasm and yet have to load the interpreted DLL too.

this it truly disappointing... I am truly regret didn't build my site using golang version of wasm

**Comment by sorainnosia** [comment]
@legistek no matter how fast the AOT performance is, when people open the site and the site does not load under 5 seconds because of downloading the big dotnet.wasm + have to download the non wasm version, user will close the tab and open competitor website instead

The browser has no chance for running the high performance wasm

**Comment by sorainnosia** [comment]
@kenchan97, I was confident at first, and promoting it to my peers, colleagues and even build a site out of it

After seeing how it turns out, I will start learning React

**Comment by x6060x** [comment]
I'm far from an expert on the topic, but as far as I understand there are multiple wasm-related issues:

- Currently there's no wasm native support for try/catch/when/finally preventing Blazor from going full AOT (causing the performance to be non-optimal)
- Because we can't go full-AOT, that means that we can't have full assembly trimming, meaning that the full original dll-s still need to be sent to the client and interpreted when needed.
- Currently there's no wasm support for DOM-manipulation - so in order to manipulate the DOM, JS interop needs to be used for that
- ...more (such as no wasm multi-threading support in all major browsers)

So my "wild" idea / question would be - is it possible / doable a Blazor-to-JS full AOT compiler to be created, which might solve the first 3 points from my list? Blazor-to-JS interpreted doesn't make sense IMO, because it will be basically do the same thing as what wasm is doing, but with lower speed. If we have full AOT we're going to basically write C# and execure JS after compilation. I tried searching about this idea, but couldn't find anything. Any opinions?

**Comment by andriysavin** [comment]
@x6060x you're forgetting about CLR which is compiled to wasm and is run almost as native code. .NET code can't just switch from CLR to JS runtime environment.

**Comment by zh4ui** [comment]
It seems wasm exception support is a work-in-progress <https://www.chromestatus.com/feature/4756734233018368>.

**Comment by sorainnosia** [comment]
> > It seems wasm exception support is a work-in-progress https://www.chromestatus.com/feature/4756734233018368.
> 
> The V8 blog states WebAssembly Exception Handling is shipped as part of [version 9.5](https://v8.dev/blog/v8-release-95#exception-handling) from 21 September 2021, so browsers might get this sooner than expected.

since it is supported now, it is still possible for .NET 6 to include full wasm AOT feature?

**Comment by charlesroddie** [comment]
This is the main issue tracking AOT support in Blazor so should be reopened.
